### PR TITLE
[WIP] Generic Select for IO

### PR DESCRIPTION
**Pull request by RX14**
See extended commit messages for more details.

This is WIP, as some things like selecting read/write from a single IO, and selecting on `sleep` are not yet implemented. I wanted to get this up for discussion early as there may be a better way to implement this, especially resume tokens.

### CODE REVIEW

File: src/select.cr [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    end
+  end
+
+  def self.select(*actions : Action)
+    self.select(actions)
+  end
+
+  def self.select(actions : Tuple | Array, has_else = false)
+    pollfds = Array(LibC::Pollfd).new
+    fdmap = Hash(Int32, Int32).new # Map of poll file descriptor to action index
```
Comment (RX14) [inline src/select.cr]: I would like to get rid of these `malloc` calls in `select`. Unfortunately it seems that allocating variable size arrays on the stack isn't possible.

File: spec/std/select_spec.cr [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+describe Select do
+  it "selects file descriptors" do
+    IO.pipe do |reader, writer|
+      writer.write_byte 10_u8
+
+      select
+      when byte = reader.read_byte
+        byte.should eq(10_u8)
+      end
```
Comment (Papierkorb) [inline spec/std/select_spec.cr]: Will not fail if line 10 is never reached

File: src/select.cr (hunk 1 above)
Comment (Papierkorb) [inline src/select.cr]: Maybe one could recycle the `pollfds` and `fdmap` variables? Chances are good, that a Fiber is always running the same `select` in a loop, with the same arguments, thus the sizes of both would never change and no sub-sequent allocations will be done. That, or we need to introduce VLAs :)

**Comment by RX14** [comment]
Closing because out of date and as per conversation [here](https://github.com/crystal-lang/crystal/commit/d87b34f82fef2749352299866923f86d76c2fde4#commitcomment-22034699).

**Pull request closed without merging**